src/leon3_serial_ccsds/leon3_serial_ccsds.c

```diff
@@ -0,0 +1,265 @@
+
+#include <EscaperInternal.h>
+
+#define MAX_TLS_SIZE RTEMS_ALIGN_UP( 0, RTEMS_TASK_STORAGE_ALIGNMENT )
+#define LEON3_SERIAL_CCSDS_POLL_STACK_SIZE (128 > RTEMS_MINIMUM_STACK_SIZE ?  128 : RTEMS_MINIMUM_STACK_SIZE)
+
+rtems_id brokerLock;
+
+static rtems_id leon3SerialCcsdsPoll_TCB = {0};
+
+RTEMS_ALIGNED(RTEMS_TASK_STORAGE_ALIGNMENT) 
+    static char leon3SerialCcsdsPoll_TaskBuffer[RTEMS_TASK_STORAGE_SIZE(
+    LEON3_SERIAL_CCSDS_POLL_STACK_SIZE + MAX_TLS_SIZE, RTEMS_FLOATING_POINT)];
+
+BYTE_FIFO_CREATE(uartFifoTx, Serial_CCSDS_LEON3_DECODED_PACKET_MAX_SIZE);
+BYTE_FIFO_CREATE(uartFifoRx, Serial_CCSDS_LEON3_ENCODED_PACKET_MAX_SIZE);
+
+static void UartRxCallback(volatile void *private_data) {
+  leon3_serial_ccsds_private_data *self =
+      (leon3_serial_ccsds_private_data *)private_data;
+
+  rtems_semaphore_release(self->semRx);
+}
+
+static void UartTxCallback(volatile void *private_data) {
+  leon3_serial_ccsds_private_data *self =
+      (leon3_serial_ccsds_private_data *)private_data;
+
+  rtems_semaphore_release(self->semTx);
+}
+
+static inline Uart_Id getUartId(const Serial_CCSDS_Leon3_Device_T dev)
+{
+  switch (dev) {
+    case uart0: {
+      return Uart_Id_0;
+      break;
+    }
+    case uart1: {
+      return Uart_Id_1;
+      break;
+    }
+    case uart2: {
+      return Uart_Id_2;
+      break;
+    }
+    case uart3: {
+      return Uart_Id_3;
+      break;
+    }
+    case uart4: {
+      return Uart_Id_4;
+      break;
+    }
+    case uart5: {
+      return Uart_Id_5;
+      break;
+    }
+    default: {
+      return Uart_Id_Invalid;
+    }
+  }
+}
+
+static inline Uart_Parity getUartParity(
+    const Serial_CCSDS_Leon3_Conf_T *const deviceConfiguration)
+{
+  if (deviceConfiguration->use_paritybit) {
+    if ( deviceConfiguration->parity == Serial_CCSDS_Leon3_Parity_T_odd) {
+      return Uart_Parity_Odd;
+    } else {
+      return Uart_Parity_Even;
+    }
+  } else {
+    return Uart_Parity_None;
+  }
+}
+
+static inline void Leon3SerialCcsdsInitUartInit(
+    leon3_serial_ccsds_private_data *const self,
+    const Serial_CCSDS_Leon3_Conf_T *const deviceConfiguration) {
+
+    Uart_Id id = getUartId(deviceConfiguration->devname);
+    Uart_Config config = (Uart_Config){ 0 };
+    config.isTxEnabled = true;
+    config.isRxEnabled = true;
+    config.isLoopbackModeEnabled = false;
+    config.parity = getUartParity(deviceConfiguration);
+
+    self->txHandler.callback = UartTxCallback;
+    self->txHandler.arg = self;
+
+    self->rxHandler.lengthCallback = UartRxCallback;
+    self->rxHandler.characterCallback = UartRxCallback;
+    self->rxHandler.characterArg = self;
+    self->rxHandler.targetCharacter = STOP_BYTE;
+    self->rxHandler.targetLength = Serial_CCSDS_LEON3_DECODED_PACKET_MAX_SIZE;
+    self->rxHandler.lengthArg = self;
+    
+    self->fifoTx = &uartFifoTx;
+    self->fifoRx = &uartFifoRx;
+
+    Uart_init(id, &self->uart);
+    Uart_setConfig(&self->uart, &config);
+    Uart_startup(&self->uart);
```

Comment:
Wouldn't it be safer to move the Uart_startup after the threads, semaphores and escaper are initialized?

Reply:
ack